mempool/src/pool.rs

```diff
@@ -0,0 +1,2769 @@
+            rolling_fee_rate
+        })
+    }
+    //
+
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), Error> {
+        let conflicts = self.validate_transaction(&tx)?;
+        self.store.drop_conflicts(conflicts);
+        self.finalize_tx(tx)?;
+        Ok(())
+    }
+
+    fn get_all(&self) -> Vec<&Transaction> {
+        self.store
+            .txs_by_descendant_score
+            .values()
+            .flatten()
+            .map(|id| &self.store.get_entry(id).expect("entry").tx)
+            .collect()
+    }
+
+    fn contains_transaction(&self, tx_id: &Id<Transaction>) -> bool {
+        self.store.txs_by_id.contains_key(&tx_id.get())
+    }
+
+    // TODO Consider returning an error
+    fn drop_transaction(&mut self, tx_id: &Id<Transaction>) {
+        self.store.remove_tx(tx_id);
+    }
+}
+
+fn has_duplicate_entry<T>(iter: T) -> bool
+where
+    T: IntoIterator,
+    T::Item: Ord,
+{
+    let mut uniq = BTreeSet::new();
+    iter.into_iter().any(move |x| !uniq.insert(x))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use common::address::Address;
+    use common::chain::config::create_mainnet;
+    use common::chain::transaction::{Destination, TxInput, TxOutput};
+    use common::chain::OutPointSourceId;
+    use core::panic;
+    use std::sync::atomic::{AtomicI64, Ordering};
+    use std::sync::Arc;
+
+    const DUMMY_WITNESS_MSG: &[u8] = b"dummy_witness_msg";
+
+    #[derive(Debug, PartialEq, Eq, Clone)]
+    struct ValuedOutPoint {
+        outpoint: OutPoint,
+        value: Amount,
+    }
+
+    impl std::cmp::PartialOrd for ValuedOutPoint {
+        fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+            other.value.partial_cmp(&self.value)
+        }
+    }
+
+    impl std::cmp::Ord for ValuedOutPoint {
+        fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+            other.value.cmp(&self.value)
+        }
+    }
+
+    fn dummy_input() -> TxInput {
+        let outpoint_source_id = OutPointSourceId::Transaction(Id::new(&H256::zero()));
+        let output_index = 0;
+        let witness = DUMMY_WITNESS_MSG.to_vec();
+        TxInput::new(outpoint_source_id, output_index, witness)
+    }
+
+    fn dummy_output() -> TxOutput {
+        let value = Amount::from(0);
+        let destination = Destination::PublicKey;
+        TxOutput::new(value, destination)
+    }
+
+    fn estimate_tx_size(num_inputs: usize, num_outputs: usize) -> usize {
+        let inputs = (0..num_inputs).into_iter().map(|_| dummy_input()).collect();
+        let outputs = (0..num_outputs).into_iter().map(|_| dummy_output()).collect();
+        let flags = 0;
+        let locktime = 0;
+        let size = Transaction::new(flags, inputs, outputs, locktime).unwrap().encoded_size();
+        // Take twice the encoded size of the dummy tx.Real Txs are larger than these dummy ones,
+        // but taking 3 times the size seems to work
+        let result = 3 * size;
```

Comment:
What does it mean that it seems to work?

Reply:
It means that prescribing the fee based on this size estimate ensures that txs will not fail the validation in `pays_minimum_relay_fees`. I've updated the comment to be more specific.